RecoBTag/FeatureTools/src/SeedingTracksConverter.cc

```diff
@@ -0,0 +1,231 @@
+                        masses.push_back(track->mass());
+                                                
+                    }
+                    
+                }
+                
+            }
+                
+            std::multimap<double,std::pair<btagbtvdeep::SeedingTrackInfoBuilder,std::vector<btagbtvdeep::TrackPairFeatures>>> sortedSeedsMap;
+            std::multimap<double,btagbtvdeep::TrackPairInfoBuilder> sortedNeighboursMap;
+            
+            std::vector<btagbtvdeep::TrackPairFeatures> tp_features_vector;            
+            
+            sortedSeedsMap.clear();
+            seedingT_features_vector.clear();
+            
+            unsigned int selTrackCount=0;
+
+            for(auto const& it : selectedTracks){
+                
+                selTrackCount+=1;
+                sortedNeighboursMap.clear();
+                tp_features_vector.clear();        
+                
+                if (reco::deltaR(it.track(), jet) > 0.4) continue;
+                
+                std::pair<bool,Measurement1D> ip = IPTools::absoluteImpactParameter3D(it, pv);
+                std::pair<double, Measurement1D> jet_dist =IPTools::jetTrackDistance(it, jetdirection, pv); 
+                TrajectoryStateOnSurface closest = IPTools::closestApproachToJet(it.impactPointState(),pv, jetdirection,it.field());
+                float length=999;
+                if (closest.isValid()) length=(closest.globalPosition() - pvp).mag();
+                
+                
+                if (!(ip.first && ip.second.value() >= 0.0 &&
+                      ip.second.significance() >= 1.0 && 
+                      ip.second.value() <= 9999. &&
+                      ip.second.significance() <= 9999. &&
+                      it.track().normalizedChi2()<5. &&
+                      std::fabs(it.track().dxy(pv.position())) < 2 &&
+                      std::fabs(it.track().dz(pv.position())) < 17 &&
+                      jet_dist.second.value()<0.07 &&
+                      length<5. )) 
+                    continue;                
+                
+                
+                btagbtvdeep::SeedingTrackInfoBuilder seedInfo;
+                seedInfo.buildSeedingTrackInfo(&(it), pv, jet, masses[selTrackCount-1], ip, jet_dist.second.value(), length, probabilityEstimator, computeProbabilities);
+
+                unsigned int neighbourTrackCount=0;
+                
+                for(auto const& tt : selectedTracks){
+                    
+                    neighbourTrackCount+=1;
+
+                    if(neighbourTrackCount==selTrackCount) continue;
+                    if(std::fabs(pv.z()-tt.track().vz())>0.1) continue;
+
+                    std::pair<bool,Measurement1D> t_ip = IPTools::absoluteImpactParameter3D(tt,pv);
+                    std::pair<bool,Measurement1D> t_ip2d = IPTools::absoluteTransverseImpactParameter(tt,pv);
```

Comment:
these are still called from a nested loop over the same selectedTracks.
These calls still dominate the total of the DeepFlavourTagInfoProducer.
Please define a vector of ```std::pair<bool,Measurement1D> ``` for the 3D and 2D IPs and also a vector of bools to indicate if these are filled. Then call the ```IPTools::absolute\*``` methods only when necessary.